Issue 64 date picker

CaptainGlobers opened the pull request.

kvladiimiir committed:
Нехватает инкремента даты и наложения класса при клике на дату.

kvladiimiir committed:
небольшие изменения в date-picker с ошибкой, проблема с получением даты из addDay

CaptainGlobers commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,132 @@
+
+  public month: number = this.selectDate.getMonth();
+
+  public getMonth(month): string {
```
Функционал работы с датами лучше вынести в отдельный класс например DateHelper.ts

CaptainGlobers replied:
лучше что бы функция принимала в качестве аргумента Date

kvladiimiir replied:
сделал

CaptainGlobers commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,132 @@
+    }
+  }
+
+  public getWeekDay(weekDay): string {
```
Функционал работы с датами лучше вынести в отдельный класс например DateHelper.ts

CaptainGlobers replied:
лучше что бы функция принимала в качестве аргумента Date

kvladiimiir replied:
сделал

CaptainGlobers commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,132 @@
+    }
+  }
+
+  public count: number = 1;
```
для чего count?

kvladiimiir replied:
удалил, так как теперь есть i-ый элемент и с ним удобнее работать

CaptainGlobers commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,132 @@
+
+  public count: number = 1;
+
+  public addDay(day: Date, counter: number): void {
```
данную функцию я скинул для примера, в твоем случае лучше использовать немного другую реализацию
например:   
  public addDay(day: Date, incrementDays: number): Date {
    return new Date(day.getFullYear(), day.getMonth(), day.getDate() + incrementDays);
  }

kvladiimiir replied:
сделал

CaptainGlobers commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,132 @@
+
+  public weekDay: number = this.dateNow.getDay();
+
+  public weekDays = [this.weekDay, this.weekDay + 1, this.weekDay + 2, this.weekDay + 3, this.weekDay + 4, this.weekDay + 5];
```
в данном случае думаю стоит сразу использовать массив дней:
  private fillWeekDays(): Date[] {
    const result: Date[] = [];
    const now: Date = new Date;
    for(let i: number = 0; i < 7; i++) {
      const weekDay: Date = this.addDay(now, i);
      result.push(weekDay);
    }
    
    return result;
  }

  public weekDays: Date[] = this.fillWeekDays();

kvladiimiir replied:
сделал

kvladiimiir committed:
Сделал вывод через дату а не через день недели

kvladiimiir committed:
Теперь массив принимает даты, а не дни недели

Теперь есть 3 функции которые возвращают номер дня, день недели, месяц

kvladiimiir committed:
Вынес функции работы с датами в отдельный класс WorkDate.ts

kvladiimiir committed:
Добавил класс WorkDate.ts для работы с датами

CaptainGlobers commented on FoodService/ClientApp/src/app/counter/counter.component.html:
```diff
@@ -1,5 +1,10 @@
 <h1>Counter</h1>
 
+<div>
+  <app-date-picker></app-date-picker>
```
тут не хватает эмитера как тут в 16 строчке (counter.component.html)
<<... (dateSelected)="deselectDate($event)" ...>>
пример реализации есть в DishColumns.Component.ts

kvladiimiir committed:
добавлены стили для компонента DatePicker

kvladiimiir committed:
добавлена работа со стилями и с emit

kvladiimiir committed:
убрано лишнее со страницы counter
и доработан event

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,47 @@
+})
+
+export class DatePickerComponent {
+  public workDate: WorkDate = new WorkDate();
```
Свойство использутеся только в UI самого контрола, поэтому необходимо модификтор protected вместо public

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,47 @@
+    return day.getMonth();
+  }
+
+  public weekDays: Date[] = this.workDate.fillWeekDays();
```
1) protected
2) Согласно стилям кодирования необходимо публичные свойства (переменные) помещать выше чем методы, т.е. нужно перенести свойство на строку 12

Qmannn replied:
Это поможет сгруппировать вместе методы и свойства для лучшей читаемости

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  protected getStyle(day: Date): string {
+    return day == this.selectedDate ? 'selected-date' : '';
```
в ts необходимо использовтаь знак строгого равенства ===

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+    return result;
+  }
+
+  public getMonthString(month): string {
```
Давай вместо цифр заведем enum с названиями всех месяцев
```
public enum Month {
    January = 0,
    February = 1,
    ...
}
```

Qmannn replied:
По факту из UI в метод будет передаваться все та же цифра, но в коде будет выглядеть как enum/
Так повышается читаемость

kvladiimiir replied:
сделал

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  public weekDayString(dayInWeek): string {
+    switch (dayInWeek) {
```
По аналогии с месяцами лучше добавить Enum.

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/app.module.ts:
```diff
@@ -32,6 +34,8 @@ import { DishesColumnComponent } from './Components/DishesColumn/DishesColumn.Co
       { path: '', component: HomeComponent, pathMatch: 'full' },
       { path: 'counter', component: CounterComponent },
       { path: 'fetch-data', component: FetchDataComponent },
+      { path: 'dish-card', component: DishSimpleComponent },
```
DishSimpleComponent - такого компонента нет, тут должен быть DishCardComponent наверно

kvladiimiir replied:
да, тут была ошибка, исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+    }
+  }
+
+  public weekDayString(dayInWeek): string {
```
не указан тип входного параметра

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+    return result;
+  }
+
+  public getMonthString(month): string {
```
Не указан тип входного параметра

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+  public fillWeekDays(): Date[] {
+    const result: Date[] = [];
+    const now: Date = new Date;
+    for (let i: number = 0; i < 6; i++) {
```
6 выглядит как магическое число. заведи const weekDaysCount

kvladiimiir replied:
сделал

kvladiimiir replied:
только в коммитах про это написать забыл

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,108 @@
+    return new Date(day.getFullYear(), day.getMonth(), day.getDate() + incrementDays);
+  }
+
+  public fillWeekDays(): Date[] {
```
fillWeekDays -> getWeekDays

kvladiimiir replied:
сделал

kvladiimiir committed:
1. Добавил Enum для упрощения читаемости кода, enum для month и weekDays
2. Исправил на DishCardComponent в app.module.ts
3. Теперь в WorfDate работаю с enum файлами, и поменял название fillWeekDays на getWeekDays
4. В DatePicker.Component.ts сделал классы protected вместо public, weekdays перенес выше, переделал знак равенства на знак строгого равенства

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,47 @@
+  public dateSelected: EventEmitter<Date> = new EventEmitter<Date>();
+  protected selectedDate: Date;
+
+  public monthString(numMonth: number): string {
```
Входной параметр должен быть типа MonthNumber вместо number

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DatePicker.Component.ts:
```diff
@@ -0,0 +1,47 @@
+    return this.workDate.getMonthString(numMonth);
+  }
+
+  public weekDayString(numWeekDay: number): string {
```
Так же типа нужно сделать WeekDayNumber

kvladiimiir replied:
исправил

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/Enums/MonthNumber.ts:
```diff
@@ -0,0 +1,14 @@
+export enum MonthNumber {
```
Енумы в отдельную папку выносить не нужно, расположи в той же папке где лежит компонент

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/WorkDate.ts:
```diff
@@ -0,0 +1,112 @@
+    const now: Date = new Date;
+    const weekDaysCount: number = 6;
+    for (let i: number = 0; i < weekDaysCount; i++) {
+      const weekDay: Date = this.addDay(now, i);
```
Неделя сейчас начинается с воскресенья, нужно чтобы с понедельника начиналась

kvladiimiir replied:
мне значит нужно написать условие? тип если у нас воскресенье, то начинать с понедельника? или вообще следует сделать условие, чтобы воскресенья не было?

Qmannn replied:
В общем, нужно чтобы дата всегда начиналась с понедельника: рабочая неделя. Допустим сегодня четверг 4 число, значит должна быть отображена неделя с понедельника 1 числа до субботы 6 числа, а выделанная дата должны быть текущей

kvladiimiir replied:
я понял, просто мне сначала по другому объяснили, что наша дата сегодняшняя должна быть первой

kvladiimiir replied:
сделал функцию сравнения дат

kvladiimiir committed:
переименовал workDate на dateHelper

вытащил enum из папки

kvladiimiir committed:
переименовано на dateHelper
использована функция isEqualDate

kvladiimiir committed:
правка в DateHelper.ts по ts lint

Qmannn commented on FoodService/ClientApp/src/app/Components/DatePicker/DateHelper.ts:
```diff
@@ -0,0 +1,124 @@
+    return date1.getFullYear() === date2.getFullYear() && date1.getMonth() === date2.getMonth() && date1.getDate() === date2.getDate();
+  }
+
+  public getWeekDays(): Date[] {
```
Будет лучше если в параметры этого метода передавать Date тогда метод станет универсальным и позволит строить неделю не только для текущей даты, но и для любой другой.
По факту нужно заменить now на параметр функции.

kvladiimiir replied:
сделал

kvladiimiir committed:
теперь getWeekDays функция принимает дату

kvladiimiir committed:
добавил constructor
